In [3]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns

In [4]:
df = pd.read_csv("C:/Users/spsha/Downloads/IMDB-Movie-Data.csv")

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
dtypes: float64(3), int64(4), object(5)
memory usage: 93.9+ KB


In [6]:
df['Genre'].unique()

array(['Action,Adventure,Sci-Fi', 'Adventure,Mystery,Sci-Fi',
       'Horror,Thriller', 'Animation,Comedy,Family',
       'Action,Adventure,Fantasy', 'Comedy,Drama,Music', 'Comedy',
       'Action,Adventure,Biography', 'Adventure,Drama,Romance',
       'Adventure,Family,Fantasy', 'Biography,Drama,History',
       'Animation,Adventure,Comedy', 'Action,Comedy,Drama',
       'Action,Thriller', 'Biography,Drama', 'Drama,Mystery,Sci-Fi',
       'Adventure,Drama,Thriller', 'Drama', 'Crime,Drama,Horror',
       'Action,Adventure,Drama', 'Drama,Thriller',
       'Action,Adventure,Comedy', 'Action,Horror,Sci-Fi',
       'Adventure,Drama,Sci-Fi', 'Action,Adventure,Western',
       'Comedy,Drama', 'Horror', 'Adventure,Drama,Fantasy',
       'Action,Crime,Thriller', 'Action,Crime,Drama',
       'Adventure,Drama,History', 'Crime,Horror,Thriller',
       'Drama,Romance', 'Comedy,Drama,Romance', 'Horror,Mystery,Thriller',
       'Crime,Drama,Mystery', 'Drama,Romance,Thriller',
       'Drama,History,T

In [7]:
unique_genres = []
for a in df['Genre']:
    unique_genres.extend(a.split(','))
    
set(unique_genres)

{'Action',
 'Adventure',
 'Animation',
 'Biography',
 'Comedy',
 'Crime',
 'Drama',
 'Family',
 'Fantasy',
 'History',
 'Horror',
 'Music',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Sport',
 'Thriller',
 'War',
 'Western'}

In [8]:
len(set(unique_genres))

20

In [9]:
genres_split = df['Genre'].str.get_dummies(sep =',')
df = pd.concat([df,genres_split],axis=1)
df

Rank                    Title                     Genre  \
0       1  Guardians of the Galaxy   Action,Adventure,Sci-Fi   
1       2               Prometheus  Adventure,Mystery,Sci-Fi   
2       3                    Split           Horror,Thriller   
3       4                     Sing   Animation,Comedy,Family   
4       5            Suicide Squad  Action,Adventure,Fantasy   
..    ...                      ...                       ...   
995   996     Secret in Their Eyes       Crime,Drama,Mystery   
996   997          Hostel: Part II                    Horror   
997   998   Step Up 2: The Streets       Drama,Music,Romance   
998   999             Search Party          Adventure,Comedy   
999  1000               Nine Lives     Comedy,Family,Fantasy   

                                           Description              Director  \
0    A group of intergalactic criminals are forced ...            James Gunn   
1    Following clues to the origin of mankind, a te...          Ridley Scott   
2    Three girls are kidnapped by a man with a diag...    M. Night Shyamalan   
3    In a city of humanoid animals, a hustling thea...  Christophe Lourdelet   
4    A secret government agency recruits some of th...            David Ayer   
..                                                 ...                   ...   
995  A tight-knit team of rising investigators, alo...             Billy Ray   
996  Three American college students studying abroa...              Eli Roth   
997  Romantic sparks occur between two dance studen...            Jon M. Chu   
998  A pair of friends embark on a mission to reuni...        Scot Armstrong   
999  A stuffy businessman finds himself trapped ins...      Barry Sonnenfeld   

                                                Actors  Year  \
0    Chris Pratt, Vin Diesel, Bradley Cooper, Zoe S...  2014   
1    Noomi Rapace, Logan Marshall-Green, Michael Fa...  2012   
2    James McAvoy, Anya Taylor-Joy, Haley Lu Richar...  2016   
3    Matthew McConaughey,Reese Witherspoon, Seth Ma...  2016   
4    Will Smith, Jared Leto, Margot Robbie, Viola D...  2016   
..                                                 ...   ...   
995  Chiwetel Ejiofor, Nicole Kidman, Julia Roberts...  2015   
996  Lauren German, Heather Matarazzo, Bijou Philli...  2007   
997  Robert Hoffman, Briana Evigan, Cassie Ventura,...  2008   
998  Adam Pally, T.J. Miller, Thomas Middleditch,Sh...  2014   
999  Kevin Spacey, Jennifer Garner, Robbie Amell,Ch...  2016   

     Runtime (Minutes)  Rating   Votes  ...  Horror  Music  Musical  Mystery  \
0                  121     8.1  757074  ...       0      0        0        0   
1                  124     7.0  485820  ...       0      0        0        1   
2                  117     7.3  157606  ...       1      0        0        0   
3                  108     7.2   60545  ...       0      0        0        0   
4                  123     6.2  393727  ...       0      0        0        0   
..                 ...     ...     ...  ...     ...    ...      ...      ...   
995                111     6.2   27585  ...       0      0        0        1   
996                 94     5.5   73152  ...       1      0        0        0   
997                 98     6.2   70699  ...       0      1        0        0   
998                 93     5.6    4881  ...       0      0        0        0   
999                 87     5.3   12435  ...       0      0        0        0   

     Romance  Sci-Fi  Sport  Thriller  War  Western  
0          0       1      0         0    0        0  
1          0       1      0         0    0        0  
2          0       0      0         1    0        0  
3          0       0      0         0    0        0  
4          0       0      0         0    0        0  
..       ...     ...    ...       ...  ...      ...  
995        0       0      0         0    0        0  
996        0       0      0         0    0        0  
997        1       0      0         0    0        0  
998  

In [10]:
# Shape of Dataset - (Num of Rows and Num of Columns)
df.shape
print("Number of Rows:",df.shape[0])
print("Number of Columns:",df.shape[1])

Number of Rows: 1000
Number of Columns: 32


In [11]:
# To ge all information about Dataset 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 32 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Rank                1000 non-null   int64  
 1   Title               1000 non-null   object 
 2   Genre               1000 non-null   object 
 3   Description         1000 non-null   object 
 4   Director            1000 non-null   object 
 5   Actors              1000 non-null   object 
 6   Year                1000 non-null   int64  
 7   Runtime (Minutes)   1000 non-null   int64  
 8   Rating              1000 non-null   float64
 9   Votes               1000 non-null   int64  
 10  Revenue (Millions)  872 non-null    float64
 11  Metascore           936 non-null    float64
 12  Action              1000 non-null   int64  
 13  Adventure           1000 non-null   int64  
 14  Animation           1000 non-null   int64  
 15  Biography           1000 non-null   int64  
 16  Comedy 

# Checking Missing values in dataset

In [12]:
print("Any missing valuse?", df.isnull().values.any())

Any missing valuse? True


In [13]:
df.isnull().sum()

Rank                    0
Title                   0
Genre                   0
Description             0
Director                0
Actors                  0
Year                    0
Runtime (Minutes)       0
Rating                  0
Votes                   0
Revenue (Millions)    128
Metascore              64
Action                  0
Adventure               0
Animation               0
Biography               0
Comedy                  0
Crime                   0
Drama                   0
Family                  0
Fantasy                 0
History                 0
Horror                  0
Music                   0
Musical                 0
Mystery                 0
Romance                 0
Sci-Fi                  0
Sport                   0
Thriller                0
War                     0
Western                 0
dtype: int64

In [14]:
# To get missing value precentage 
per_missing = df.isnull().sum()*100/len(df)
print(per_missing)

Rank                   0.0
Title                  0.0
Genre                  0.0
Description            0.0
Director               0.0
Actors                 0.0
Year                   0.0
Runtime (Minutes)      0.0
Rating                 0.0
Votes                  0.0
Revenue (Millions)    12.8
Metascore              6.4
Action                 0.0
Adventure              0.0
Animation              0.0
Biography              0.0
Comedy                 0.0
Crime                  0.0
Drama                  0.0
Family                 0.0
Fantasy                0.0
History                0.0
Horror                 0.0
Music                  0.0
Musical                0.0
Mystery                0.0
Romance                0.0
Sci-Fi                 0.0
Sport                  0.0
Thriller               0.0
War                    0.0
Western                0.0
dtype: float64


* Here in column Revenue (Millions) 12.8% are missing values and in column Metascore 6.4% missing values are there.

In [15]:
# Check for duplicate Data
duplicate_data = df.duplicated().any()
print(duplicate_data)

False


* To fill missing values 

In [16]:
# taking mean value of whole Revenue (Millions) column 
a=df['Revenue (Millions)'].mean()
a

82.95637614678898

* This value is mean of Revenue (Millions) column in dataset

In [17]:
# To fill this mean in missing values
df['Revenue (Millions)'].fillna(a,inplace=True)

In [18]:
# taking mean value of whole Metascore column
b=df['Metascore'].mean()
b

58.98504273504273

* This value is mean of Metascore column in dataset

In [19]:
# # To fill this mean in missing values
df['Metascore'].fillna(b,inplace=True)

In [20]:
# to check if there is any missing values 
print("Any missing valuse?", df.isnull().values.any())

Any missing valuse? False


In [21]:
# To get statistics about the dataframe 
df.describe()

Rank         Year  Runtime (Minutes)       Rating         Votes  \
count  1000.000000  1000.000000        1000.000000  1000.000000  1.000000e+03   
mean    500.500000  2012.783000         113.172000     6.723200  1.698083e+05   
std     288.819436     3.205962          18.810908     0.945429  1.887626e+05   
min       1.000000  2006.000000          66.000000     1.900000  6.100000e+01   
25%     250.750000  2010.000000         100.000000     6.200000  3.630900e+04   
50%     500.500000  2014.000000         111.000000     6.800000  1.107990e+05   
75%     750.250000  2016.000000         123.000000     7.400000  2.399098e+05   
max    1000.000000  2016.000000         191.000000     9.000000  1.791916e+06   

       Revenue (Millions)    Metascore       Action    Adventure    Animation  \
count         1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean            82.956376    58.985043     0.303000     0.259000     0.049000   
std             96.412043    16.634858     0.459785     0.438305     0.215976   
min              0.000000    11.000000     0.000000     0.000000     0.000000   
25%             17.442500    47.750000     0.000000     0.000000     0.000000   
50%             60.375000    58.985043     0.000000     0.000000     0.000000   
75%             99.177500    71.000000     1.000000     1.000000     0.000000   
max            936.630000   100.000000     1.000000     1.000000     1.000000   

       ...       Horror        Music      Musical      Mystery      Romance  \
count  ...  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000   
mean   ...     0.119000     0.016000     0.005000     0.106000     0.141000   
std    ...     0.323951     0.125538     0.070569     0.307992     0.348196   
min    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
25%    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
50%    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
75%    ...     0.000000     0.000000     0.000000     0.000000     0.000000   
max    ...     1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi        Sport     Thriller          War      Western  
count  1000.000000  1000.000000  1000.000000  1000.000000  1000.000000  
mean      0.120000     0.018000     0.195000     0.013000     0.007000  
std       0.325124     0.133018     0.396399     0.113331     0.083414  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000     1.000000  

[8 rows x 27 columns]

In [22]:
df.to_csv('data_csv',index=False)

# K-Nearest Neighbor algorithm

In [23]:
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.neighbors import KNeighborsRegressor
import numpy as np

In [24]:
#Key Points:
#Data Inspection: It is important to inspect the data types of each column to identify which columns need encoding.
#One-Hot Encoding: Convert all categorical columns to numerical using one-hot encoding.
#Handle Missing Values: Ensure all missing values are handled appropriately.
#Standardization: Only standardize numeric columns to have a mean of 0 and a standard deviation of 1.
#Troubleshooting Tips:
#Check for Additional Categorical Data: Ensure there are no other string columns that need encoding.
#Missing Values: Confirm that missing values are filled appropriately to avoid NaNs.
#Data Types: Double-check that after encoding and filling missing values, all columns are numeric.
#By following these steps, you should be able to resolve the error and successfully apply the KNN algorithm to your dataset.
#If you still encounter issues, please provide more details or a sample of the dataset for further troubleshooting.

In [25]:
# Feature selection (exclude categorical and text features for simplicity)
features = df[['Year', 'Runtime (Minutes)', 'Rating', 'Votes', 'Rank', 'Metascore', 'Action', 'Adventure', 
               'Animation', 'Biography', 'Comedy', 'Crime', 'Drama', 'Family', 'Fantasy', 'History', 'Horror', 'Music', 
               'Musical', 'Mystery', 'Romance', 'Sci-Fi', 'Sport', 'Thriller', 'War', 'Western']]
target =np.log1p(df['Revenue (Millions)'])

# Split the data
X_train, X_test, Y_train, Y_test = train_test_split(features, target, test_size=0.2, random_state=42)

# Standardize the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Train a kNN Regression model
k = 7  # Number of neighbors
knn_model = KNeighborsRegressor(n_neighbors=k)
knn_model.fit(X_train_scaled, Y_train)

Y_pred = knn_model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f'Mean Squared Error: {mse}')
rmse= np.sqrt(mse)
print(f'Root Mean Squared Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Mean Squared Error: 2.032153295805875
Root Mean Squared Error: 1.4255361432828966
R Squared: 0.19578459728415598


In [26]:
# Define the kNN model
knn = KNeighborsRegressor()

# Define the grid of hyperparameters
param_grid = {
    'n_neighbors': [3, 5, 7, 9, 11],
    'weights': ['uniform', 'distance'],
    'metric': ['euclidean', 'manhattan', 'minkowski']
}

# Perform grid search
grid_search = GridSearchCV(knn, param_grid, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(X_train_scaled, Y_train)

# Get the best parameters and best model
best_params = grid_search.best_params_
best_model = grid_search.best_estimator_

print(f"Best Parameters: {best_params}")

# Predict and evaluate
Y_pred = best_model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f'Mean Squared Error: {mse}')
rmse= np.sqrt(mse)
print(f'Root Mean Squared Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Best Parameters: {'metric': 'euclidean', 'n_neighbors': 11, 'weights': 'distance'}
Mean Squared Error: 1.9535756969706666
Root Mean Squared Error: 1.3977037228864586
R Squared: 0.2268813238067685


# Neural Network

In [27]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from keras.models import Sequential
from keras.layers import Dense, Activation
import numpy as np

In [28]:
# Define the neural network model
model = Sequential()
model.add(Dense(64, activation='relu', input_shape=(X_train_scaled.shape[1],)))
model.add(Dense(32, activation='relu'))
model.add(Dense(1))  # Output layer for regression

# Compile the model
model.compile(optimizer='adam', loss='mean_squared_error')

# Train the model
model.fit(X_train_scaled, Y_train, epochs=50, batch_size=10, validation_split=0.2)

# Predict and evaluate
Y_pred = model.predict(X_test_scaled)
mse = mean_squared_error(Y_test, Y_pred)
print(f'Mean Squared Error: {mse}')
rmse= np.sqrt(mse)
print(f'Root Mean Squared Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

C:\Users\spsha\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Epoch 1/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 3s 7ms/step - loss: 9.0304 - val_loss: 2.4516
Epoch 2/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 2.1614 - val_loss: 1.8691
Epoch 3/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 1.6245 - val_loss: 1.7466
Epoch 4/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - loss: 1.6153 - val_loss: 1.7345
Epoch 5/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.4794 - val_loss: 1.6976
Epoch 6/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.4599 - val_loss: 1.7057
Epoch 7/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.4861 - val_loss: 1.6926
Epoch 8/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.2300 - val_loss: 1.6975
Epoch 9/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.2365 - val_loss: 1.7749
Epoch 10/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - loss: 1.2923 - val_loss: 1.6676
Epoch 11/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.0789 - val_loss: 1.6632
Epoch 12/50
64/64 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - loss: 1.1114 - val_lo

In [29]:
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense

# Create the neural network model
model = Sequential()
model.add(Dense(64, input_dim=X_train.shape[1], activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(1))  # Output layer for regression

model.compile(optimizer='adam', loss='mean_squared_error')

model.fit(X_train, Y_train, epochs=50, batch_size=32, validation_split=0.2)

Y_pred = model.predict(X_test)
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Epoch 1/50


C:\Users\spsha\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - loss: 145117520.0000 - val_loss: 46699480.0000
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 14486807.0000 - val_loss: 6030671.0000
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 1967139.6250 - val_loss: 11597.4736
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 23195.6016 - val_loss: 20210.6582
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 17707.5430 - val_loss: 15827.4590
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 10039.2402 - val_loss: 5287.8662
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 2142.9294 - val_loss: 713.3235
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - loss: 578.7393 - val_loss: 1334.2367
Epoch 9/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 456.7712 - val_loss: 87.6652
Epoch 10/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 82.5336 - val_loss: 79.7516
Epoch 11/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - loss: 43.2276 - val_loss: 67.7623
Epoch 

# Bagging

In [33]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import BaggingRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error

In [36]:
# Create a Bagging Regressor
bagging_model = BaggingRegressor(base_estimator=DecisionTreeRegressor(), n_estimators=100, random_state=42)
bagging_model.fit(X_train, Y_train)

# Make predictions and evaluate the model
Y_pred = bagging_model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error with Bagging: {mse}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

C:\Users\spsha\anaconda3\Lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


Mean Squared Error with Bagging: 1.8135041014114137
Root Mean Square Error: 1.346664064052878
R Squared: 0.282314019196544


In [39]:
# Define the parameter grid
param_grid = {
    'n_estimators': [10, 50, 100],
    'max_samples': [0.5, 0.7, 1.0],
    'max_features': [0.5, 0.7, 1.0],
}

# Grid Search CV
grid_search = GridSearchCV(estimator=bagging_model, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)
grid_search.fit(X_train, Y_train)

# Best parameters and model evaluation
print("Best parameters:", grid_search.best_params_)
best_model = grid_search.best_estimator_
Y_pred = best_model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error with Bagging: {mse}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

C:\Users\spsha\anaconda3\Lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


Best parameters: {'max_features': 1.0, 'max_samples': 0.5, 'n_estimators': 100}
Mean Squared Error with Bagging: 1.733135772984643
Root Mean Square Error: 1.3164861461423143
R Squared: 0.31411941879151073


# Boosting

In [37]:
from sklearn.ensemble import GradientBoostingRegressor

# Create a Gradient Boosting Regressor
boosting_model = GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)
boosting_model.fit(X_train, Y_train)

# Make predictions and evaluate the model
Y_pred = boosting_model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error with Boosting: {mse}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Mean Squared Error with Boosting: 1.8526078002479627
Root Mean Square Error: 1.3611053597161253
R Squared: 0.2668389086463612


In [42]:
# Define the bagging model with boosting as the base estimator
bagging_boosting_model = BaggingRegressor(base_estimator=boosting_model, random_state=42)

# Define the parameter grid
param_grid = {
    'base_estimator__n_estimators': [100, 200],
    'base_estimator__learning_rate': [0.01, 0.1],
    'base_estimator__max_depth': [3, 4],
    'n_estimators': [10, 20],
    'max_samples': [0.8, 1.0],
    'max_features': [0.8, 1.0]
}

# Grid Search CV
grid_search = GridSearchCV(estimator=bagging_boosting_model, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)
grid_search.fit(X_train, Y_train)

# Best parameters and model evaluation
print("Best parameters:", grid_search.best_params_)
best_model = grid_search.best_estimator_

Y_pred = best_model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error with Boosting: {mse}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

C:\Users\spsha\anaconda3\Lib\site-packages\sklearn\ensemble\_base.py:166: FutureWarning: `base_estimator` was renamed to `estimator` in version 1.2 and will be removed in 1.4.
  warnings.warn(


Best parameters: {'base_estimator__learning_rate': 0.1, 'base_estimator__max_depth': 3, 'base_estimator__n_estimators': 100, 'max_features': 1.0, 'max_samples': 0.8, 'n_estimators': 10}
Mean Squared Error with Boosting: 1.7620351619341408
Root Mean Square Error: 1.3274167250468636
R Squared: 0.3026826173601278


# Staking 

In [49]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.ensemble import StackingRegressor
from sklearn.metrics import mean_squared_error

In [50]:
# Step 3: Define Base Models
base_models = [
    ('rf', RandomForestRegressor(n_estimators=100, random_state=42)),
    ('gb', GradientBoostingRegressor(n_estimators=100, random_state=42)),
    ('svr', SVR(kernel='rbf', C=1.0, epsilon=0.1))
]

# Step 4: Define the Meta-Model
meta_model = LinearRegression()

# Step 5: Create the Stacking Ensemble
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model)

# Step 6: Train and Evaluate the Model
stacking_model.fit(X_train, Y_train)
Y_pred = stacking_model.predict(X_test)

# Evaluate the performance
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error: {mse:.2f}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Mean Squared Error: 1.75
Root Mean Square Error: 1.3239481881347706
R Squared: 0.30632203374611944


In [51]:
# Step 3: Define Base Models with Hyperparameters for Grid Search
base_models = [
    ('rf', RandomForestRegressor(random_state=42)),
    ('gb', GradientBoostingRegressor(random_state=42)),
    ('svr', SVR())
]

# Define parameter grids for each base model
param_grid = {
    'rf__n_estimators': [50, 100],
    'rf__max_depth': [None, 10, 20],
    'gb__n_estimators': [50, 100],
    'gb__learning_rate': [0.01, 0.1],
    'svr__C': [0.1, 1.0, 10],
    'svr__epsilon': [0.1, 0.2]
}

# Step 4: Define the Meta-Model with Hyperparameters for Grid Search
meta_model = LinearRegression()

# Step 5: Create the Stacking Ensemble
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model)

# Step 6: Perform Grid Search CV
grid_search = GridSearchCV(estimator=stacking_model, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)

# Fit the grid search to the data
grid_search.fit(X_train, Y_train)

# Get the best model
best_model = grid_search.best_estimator_

# Step 7: Train and Evaluate the Best Model
Y_pred = best_model.predict(X_test)

# Evaluate the performance
mse = mean_squared_error(Y_test, Y_pred)
print(f"Best Model Mean Squared Error: {mse:.2f}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Best Model Mean Squared Error: 1.76
Root Mean Square Error: 1.3247906932931497
R Squared: 0.30543889764974474


# Voting Regressor

In [53]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.svm import SVR
from sklearn.ensemble import VotingRegressor
from sklearn.metrics import mean_squared_error

In [54]:
# Step 3: Define Base Models
base_models = [
    ('lr', LinearRegression()),
    ('rf', RandomForestRegressor(n_estimators=100, random_state=42)),
    ('gb', GradientBoostingRegressor(n_estimators=100, random_state=42)),
    ('svr', SVR(kernel='rbf', C=1.0, epsilon=0.1))
]

# Step 4: Create the Voting Regressor
voting_regressor = VotingRegressor(estimators=base_models)

# Step 5: Train and Evaluate the Model
voting_regressor.fit(X_train, Y_train)
Y_pred = voting_regressor.predict(X_test)

# Evaluate the performance
mse = mean_squared_error(Y_test, Y_pred)
print(f"Voting Regressor Mean Squared Error: {mse:.2f}")
rmse = np.sqrt(mean_squared_error(Y_test, Y_pred))
print(f'Root Mean Square Error: {rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Voting Regressor Mean Squared Error: 1.67
Root Mean Square Error: 1.2941135774494872
R Squared: 0.337233256449602


In [56]:
# Step 3: Define Base Models with Hyperparameters for Grid Search
base_models = [
    ('rf', RandomForestRegressor(random_state=42)),
    ('gb', GradientBoostingRegressor(random_state=42)),
    ('svr', SVR())
]

# Define parameter grids for each base model
param_grid = {
    'rf__n_estimators': [50, 100],
    'rf__max_depth': [None, 10, 20],
    'gb__n_estimators': [50, 100],
    'gb__learning_rate': [0.01, 0.1],
    'svr__C': [0.1, 1.0, 10],
    'svr__epsilon': [0.1, 0.2]
}

# Step 4: Create the Voting Regressor
voting_regressor = VotingRegressor(estimators=base_models)

# Step 5: Perform Grid Search CV
grid_search = GridSearchCV(estimator=voting_regressor, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', n_jobs=-1)

# Fit the grid search to the data
grid_search.fit(X_train, Y_train)

# Get the best model
best_model = grid_search.best_estimator_

# Step 6: Train and Evaluate the Best Model
Y_pred = best_model.predict(X_test)

# Evaluate the performance
mse = mean_squared_error(Y_test, Y_pred)
print(f"Best Model Mean Squared Error: {mse:.2f}")
rmse=np.sqrt(mse)
print(f'Root Mean Squared Error:{rmse}')
r2=r2_score(Y_test, Y_pred)
print(f'R Squared: {r2}')

Best Model Mean Squared Error: 1.73
Root Mean Squared Error:1.3136676811393144
R Squared: 0.31705307836813057


# linear regression

In [57]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import mean_squared_error

In [60]:
model = LinearRegression()
model.fit(X_train, Y_train)
coefficients = model.coef_
model = LinearRegression()
model.fit(X_train, Y_train)

coefficients = pd.DataFrame({'Feature': X_train.columns,'Coefficient': model.coef_})
coefficients

X_train, X_test, Y_train, Y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=0)
model = LinearRegression()
model.fit(X_train, Y_train)
Y_pred = model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error: {mse}")
rmse=np.sqrt(mse)
print(f'root Mean Squared Error: {rmse}')
rmse
r2=r2_score(Y_test,Y_pred)
print(f' R^2:{r2}')

Mean Squared Error: 1.6079643952506413
root Mean Squared Error: 1.2680553596947735
 R^2:0.28557451744205153


In [85]:
X_train, X_test, Y_train, Y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=0)
model = LinearRegression()
model.fit(X_train, Y_train)

param_grid = {
    'fit_intercept': [True, False],
    'copy_X': [True, False]
}
grid_search = GridSearchCV(estimator=model, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error', verbose=1)
grid_search.fit(X_train, Y_train)
best_model = grid_search.best_estimator_
Y_pred = best_model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
print(f"Best Model Mean Squared Error: {mse}")
best_params = grid_search.best_params_
print(f"Best Parameters: {best_params}")
mse = mean_squared_error(Y_test, Y_pred)
print(f"Mean Squared Error: {mse}")
rmse=np.sqrt(mse)
print(f'Root Mean Squared Error:{rmse}')

Fitting 5 folds for each of 4 candidates, totalling 20 fits
Best Model Mean Squared Error: 1.9737116568796753
Best Parameters: {'copy_X': True, 'fit_intercept': False}
Mean Squared Error: 1.9737116568796753
Root Mean Squared Error:1.4048884855673334


# SVR

In [63]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score

In [65]:
X_train, X_test, Y_train, Y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=0)


scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)


svr_model = SVR(kernel='rbf')  
svr_model.fit(X_train_scaled, Y_train)


Y_pred = svr_model.predict(X_test_scaled)
Y_pred

mse = mean_squared_error(Y_test, Y_pred)
print(f'Mean Squared Error: {mse}')
rmse=np.sqrt(mse)
print(f'root Mean Squared Error: {rmse}')
rmse

Mean Squared Error: 1.9219292286550265
root Mean Squared Error: 1.386336621695837


1.386336621695837

In [66]:
X_train, X_test, Y_train, Y_test = train_test_split(X_train, Y_train, test_size=0.2, random_state=0)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
svr = SVR()

param_grid = {'kernel': ['rbf'],'C': [0.1, 1, 1],'gamma': ['scale', 'auto'], 'epsilon': [0.1, 0.2, 0.5]} # kernel type: linear or radial basis function (gaussian)             # regularization parameter# kernel coefficient (for 'rbf' kernel) # epsilon parameter in SVR

grid_search = GridSearchCV(estimator=svr, param_grid=param_grid, scoring='neg_mean_squared_error', cv=5, verbose=1, n_jobs=-1)

grid_search.fit(X_train_scaled, Y_train)
print("Best parameters found: ", grid_search.best_params_)
print("Lowest RMSE found: ", np.sqrt(np.abs(grid_search.best_score_)))
best_model = grid_search.best_estimator_
Y_pred = best_model.predict(X_test_scaled)

mse = mean_squared_error(Y_test, Y_pred)
print(f'Mean Squared Error: {mse}')
rmse=np.sqrt(mse)
print(f'root Mean Squared Error: {rmse}')
rmse

Fitting 5 folds for each of 18 candidates, totalling 90 fits
Best parameters found:  {'C': 1, 'epsilon': 0.2, 'gamma': 'auto', 'kernel': 'rbf'}
Lowest RMSE found:  1.219227469341784
Mean Squared Error: 1.3262195408521018
root Mean Squared Error: 1.151616056180228


1.151616056180228

# Decision Tree Regression

In [68]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score, make_scorer
from sklearn.model_selection import GridSearchCV

In [86]:
model_dt = DecisionTreeRegressor(random_state=5)

# Train the model
model_dt.fit(X_train, Y_train)

# Make predictions
Y_pred = model_dt.predict(X_test)

# Evaluate the model
mse = mean_squared_error(Y_test, Y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(Y_test, Y_pred)

print("Decision Tree Regression:")
print(f'Mean Squared Error: {mse}')
print(f'Root Mean Squared Error: {rmse}')
print(f'R^2 Score: {r2}')

Decision Tree Regression:
Mean Squared Error: 1.6715536093339198
Root Mean Squared Error: 1.292885768091644
R^2 Score: -0.9485966996416622


In [87]:
# Define the Decision Tree Regressor model
model_dt = DecisionTreeRegressor(random_state=5)

# Define the parameter grid
param_grid = {
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Grid Search CV
grid_search = GridSearchCV(estimator=model_dt, param_grid=param_grid, cv=5, scoring='neg_mean_squared_error')
grid_search.fit(X_train, Y_train)

#best parameters 
print("Best parameters found: ", grid_search.best_params_)
print("Best negative mean squared error: ", grid_search.best_score_)

# Evaluate the best model on test data
best_model = grid_search.best_estimator_
Y_pred = best_model.predict(X_test)
mse = mean_squared_error(Y_test, Y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(Y_test, Y_pred)

print("\nBest Model Evaluation:")
print(f'Best Model Mean Squared Error: {mse}')
print(f'Best Model Root Mean Squared Error: {rmse}')
print(f'Best Model R^2 Score: {r2}')

Best parameters found:  {'max_depth': None, 'min_samples_leaf': 1, 'min_samples_split': 10}
Best negative mean squared error:  -1.6325994003010376

Best Model Evaluation:
Best Model Mean Squared Error: 1.3175554226045674
Best Model Root Mean Squared Error: 1.1478481705367516
Best Model R^2 Score: -0.5359268968377784


# Random Forest Algorithm 

* Random Forest or random decision forest is an ensemble learning method for classification, regression.
* Random decision forests correct for decision tree's habit overfitting to their trainning set.
* Step1 : Create Bootstrap Dataset from Original data by randomly choosing data(repetation is allowed)
* Step2 : Create Randomized Decision Tress from Bootstap dataset.
* Step3 : Finally output of the random foresr is the class selected by most trees.

In [88]:
from sklearn.ensemble import RandomForestRegressor

# Random Forest Regression model
model_rf = RandomForestRegressor(n_estimators=100, random_state=5)

# Train the model
model_rf.fit(X_train, Y_train)

# Make predictions
Y_pred = model_rf.predict(X_test)

# Evaluate the model
mse = mean_squared_error(Y_test, Y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(Y_test, Y_pred)

print("Random Forest Regression:")
print(f'Mean Squared Error: {mse}')
print(f'Root Mean Squared Error: {rmse}')
print(f'R^2 Score: {r2}')

Random Forest Regression:
Mean Squared Error: 1.0168950431794117
Root Mean Squared Error: 1.0084121395438532
R^2 Score: -0.18543510298240573


In [89]:
from sklearn.model_selection import GridSearchCV

# Define the RandomForestRegressor
model_rf = RandomForestRegressor(random_state=5)

# Define the parameter grid
param_grid = {
    'n_estimators': [100, 200, 300],
    'max_features': ['auto', 'sqrt', 'log2'],
    'max_depth': [None, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}


# Define the scoring methods 
scoring = {
    'MSE': make_scorer(mean_squared_error),
    'R^2': make_scorer(r2_score)
}

# Model
grid_search = GridSearchCV(estimator=model_rf, param_grid=param_grid, 
                           scoring=scoring, refit='MSE', cv=5, verbose=1, n_jobs=-1)

# To fit GridSearchCV
grid_search.fit(X_train, Y_train)

# Best parameters and best score found by GridSearchCV
print("Best parameters found by GridSearchCV:")
print(grid_search.best_params_)
print("Best score (MSE) found by GridSearchCV:")
print(grid_search.best_score_)

# Make predictions with the best estimator
best_model = grid_search.best_estimator_
Y_pred = best_model.predict(X_test)

# Evaluate the best model
mse = mean_squared_error(Y_test, Y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(Y_test, Y_pred)

# Print evaluation metrics of the best model
print("\nEvaluation metrics of the best model:")
print(f'Mean Squared Error: {mse}')
print(f'Root Mean Squared Error: {rmse}')
print(f'R^2 Score: {r2}')

Fitting 5 folds for each of 486 candidates, totalling 2430 fits


C:\Users\spsha\anaconda3\Lib\site-packages\sklearn\ensemble\_forest.py:413: FutureWarning: `max_features='auto'` has been deprecated in 1.1 and will be removed in 1.3. To keep the past behaviour, explicitly set `max_features=1.0` or remove this parameter as it is also the default value for RandomForestRegressors and ExtraTreesRegressors.
  warn(


Best parameters found by GridSearchCV:
{'max_depth': None, 'max_features': 'auto', 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100}
Best score (MSE) found by GridSearchCV:
1.4927718910565015

Evaluation metrics of the best model:
Mean Squared Error: 1.041754185580286
Root Mean Squared Error: 1.0206636005953607
R^2 Score: -0.2144143965975045


In [3]:
import pandas as pd
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier, BaggingClassifier, AdaBoostClassifier, GradientBoostingClassifier, VotingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.svm import SVR
from sklearn.metrics import accuracy_score

In [4]:
# Example, adjust parameters as needed
Data = {'model' :  ['Linear Regression','KNN','Decision Tree','Random Forest','Bagging','Neural network','Gradient Boosting','Stacking','Voting','SVR'],
        'RMSE' : ['1.2680553596947735','1.4255361432828966','1.292885768091644','1.0084121395438532','1.346664064052878','1.6023867187681156',
                  '1.3611053597161253','1.3239481881347706','1.2941135774494872','1.386336621695837'],
        'Grid Search CV' : ['1.4048884855673334','1.3977037228864586','1.1478481705367516','1.0206636005953607','1.3164861461423143',
                            '3.7594706884059854','1.3274167250468636','1.3247906932931497','1.3136676811393144','1.151616056180228']
}
print(pd.DataFrame(Data))

               model                RMSE      Grid Search CV
0  Linear Regression  1.2680553596947735  1.4048884855673334
1                KNN  1.4255361432828966  1.3977037228864586
2      Decision Tree   1.292885768091644  1.1478481705367516
3      Random Forest  1.0084121395438532  1.0206636005953607
4            Bagging   1.346664064052878  1.3164861461423143
5     Neural network  1.6023867187681156  3.7594706884059854
6  Gradient Boosting  1.3611053597161253  1.3274167250468636
7           Stacking  1.3239481881347706  1.3247906932931497
8             Voting  1.2941135774494872  1.3136676811393144
9                SVR   1.386336621695837   1.151616056180228
